src/ray/gcs/gcs_client/gcs_client.h

```diff
@@ -69,11 +69,9 @@ class RAY_EXPORT GcsClient : public std::enable_shared_from_this<GcsClient> {
   ///
   /// \param options Options for client.
   /// \param get_gcs_server_address_func Function to get GCS server address.
-  explicit GcsClient(const GcsClientOptions &options,
-                     std::function<bool(std::pair<std::string, int> *)>
-                         get_gcs_server_address_func = nullptr);
+  explicit GcsClient(const GcsClientOptions &options);
 
-  virtual ~GcsClient() = default;
+  virtual ~GcsClient() { Disconnect(); };
```

Comment:
The disconnect is to stop the io thread.

Disconnect is to give the user a way to stop the client and later do connection. I feel if this is like you said, it should be some issues related to something higher layer. Not sure whether the function is needed, but when we delete the GCS client, we need to stop something in io context and this is to avoid race condition.

Even if there is no io context issues here and I don't do disconnect here, the ordering of destruction depends on the order it's declared. So disconnect shouldn't be the root cause of the issue you mentioned. 

The root cause is that we have complicated dependence on core worker which makes everything complicated.